File: src/main/java/s4j/java/chapter12/Customer.java
package s4j.java.chapter12;

@SuppressWarnings("all")
public class Customer implements Comparable<Customer> {

    private final String name;
    private final String address;
    private final ShoppingBasket basket = new ShoppingBasket();

    public static Customer create(String name, String address, Integer yearsACustomer) {
        Customer customer = null;
        if (yearsACustomer > 2) {
            customer = new DiscountedCustomer(name, address);
        } else {
            customer = new Customer(name, address);
        }
        return customer;
    }

    public Customer(String name, String address) {
        this.name = name;
        this.address = address;
    }

    public void add(Item item) {
        basket.add(item);
    }

    public Double total() {
        return basket.value();
    }

    public String getName() {
        return name;
    }

    @Override
    public int compareTo(Customer other) {
        return name.compareTo(other.name);
    }

    @Override
    public String toString() {
        return name + " $ " + total();
    }
}

File: src/main/scala/s4j/scala/chapter12/Customer.scala
/*
 * Copyright (c) 2015-2017 Toby Weston
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package s4j.scala.chapter12

class Customer(val name: String, val address: String) extends Ordered[Customer] {

  private final val basket: ShoppingBasket = new ShoppingBasket

  def add(item: Item) {
    basket.add(item)
  }

  def total: Double = {
    basket.value
  }

  override def compare(that: Customer): Int = name.compare(that.name)

  override def toString: String = name + " $ " + total
}

package customer {

  object Example extends App {
    val customer = new Customer("", "")
    customer.add(PricedItem(2.3))
    customer.add(PricedItem(3.3))
    println("customer.total = " + customer.total)

    val discounted = new DiscountedCustomer("George", "1 Apple Street")
    discounted.add(PricedItem(2.3))
    discounted.add(PricedItem(3.3))
    println("discounted.total = " + discounted.total)
  }

}